vim-patch:8.2.{1727,2324,2363}: charcol(), getcharpos(), getcursorcharpos(), setcharpos(), setcursorcharpos()

zeertzjq opened the pull request:
#### vim-patch:8.2.1727: a popup created with "cursorline" will ignore "firstline"

Problem:    A popup created with "cursorline" will ignore "firstline".
Solution:   When both "cursorline" and "firstline" are present put the cursor
            on "firstline".  Add the "winid" argument to
            getcurpos().
https://github.com/vim/vim/commit/99ca9c4868bb1669706b9e3de9a9218bd11cc459

Skip popup window related code.
Cherry-pick all of Test_getcurpos_setpos() from patch 8.2.0610.


#### vim-patch:8.2.2324: not easy to get mark en cursor posotion by character count

Problem:    Not easy to get mark en cursor posotion by character count.
Solution:   Add functions that use character index. (Yegappan Lakshmanan,
            closes vim/vim#7648)
https://github.com/vim/vim/commit/6f02b00bb0958f70bc15534e115b4c6dadff0e06


#### vim-patch:8.2.2363: curpos() does not accept a string argument as before

Problem:    curpos() does not accept a string argument as before.
solution:   Make a string argument work again. (Yegappan Lakshmanan,
            closes vim/vim#7690
https://github.com/vim/vim/commit/9ebcf231bdccc1673cc92b20f5190fc577ad29d0


#### vim-patch:partial:2346a6378483

Update runtime files
https://github.com/vim/vim/commit/2346a6378483c9871016f9fc821ec5cbea638f13


#### vim-patch:partial:6aa57295cfbe

Update runtime files
https://github.com/vim/vim/commit/6aa57295cfbe8f21c15f0671e45fd53cf990d404

zeertzjq commented:
@bfredl Is it fine to consider the remaining code from Vim patch 8.2.1727 as N/A?

bfredl commented:
@zeertzjq yes, looks good to me.

zeertzjq commented on src/nvim/eval.c:
```diff
@@ -8151,6 +8151,58 @@ char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
+    lnum = buf->b_ml.ml_line_count;
+  }
+
+  char_u *str = ml_get_buf(buf, lnum, false);
+  if (str == NULL) {
+    return -1;
+  }
```
I cannot infer this directly, but I found that many places in Vim use the result of `ml_get_buf()` without checking if it is `NULL`.

zeertzjq replied:
Looks like such a `NULL` check never appears anywhere else, so I'll remove it.

seandewar reviewed (approved):
LGTM otherwise; let's hope that NULL check wasn't needed 😄

zeertzjq committed:
vim-patch:partial:2346a6378483

Update runtime files
https://github.com/vim/vim/commit/2346a6378483c9871016f9fc821ec5cbea638f13

zeertzjq committed:
vim-patch:8.2.2324: not easy to get mark en cursor posotion by character count

Problem:    Not easy to get mark en cursor posotion by character count.
Solution:   Add functions that use character index. (Yegappan Lakshmanan,
            closes vim/vim#7648)
https://github.com/vim/vim/commit/6f02b00bb0958f70bc15534e115b4c6dadff0e06

zeertzjq committed:
vim-patch:8.2.1727: a popup created with "cursorline" will ignore "firstline"

Problem:    A popup created with "cursorline" will ignore "firstline".
Solution:   When both "cursorline" and "firstline" are present put the cursor
            on "firstline". (closes vim/vim#7000)  Add the "winid" argument to
            getcurpos().
https://github.com/vim/vim/commit/99ca9c4868bb1669706b9e3de9a9218bd11cc459

Skip popup window related code.
Cherry-pick all of Test_getcurpos_setpos() from patch 8.2.0610.

zeertzjq committed:
vim-patch:8.2.2363: curpos() does not accept a string argument as before

Problem:    curpos() does not accept a string argument as before.
solution:   Make a string argument work again. (Yegappan Lakshmanan,
            closes vim/vim#7690
https://github.com/vim/vim/commit/9ebcf231bdccc1673cc92b20f5190fc577ad29d0

zeertzjq committed:
vim-patch:partial:6aa57295cfbe

Update runtime files
https://github.com/vim/vim/commit/6aa57295cfbe8f21c15f0671e45fd53cf990d404

zeertzjq commented on src/nvim/eval.lua:
```diff
@@ -311,8 +314,10 @@ return {
     serverstop={args=1},
     setbufline={args=3, base=3},
     setbufvar={args=3, base=3},
+    setcharpos={args={2, 2}, base=2},
```
Thanks!
